src/rest-devices.c

```diff
@@ -0,0 +1,575 @@
+            if (device_data->psk == NULL)
+            {
+                device_data->psk = oldpsk;
+                return -1;
+            }
+            memcpy(device_data->psk, binary_buffer, length);
+
+            jstring = json_object_get(jdevice, "psk_id");
+            string = json_string_value(jstring);
+
+            base64_decode(string, NULL, &length);
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                return -1;
+            }
+
+            oldid = device_data->psk_id;
+            device_data->psk_id = malloc(length);
+            if (device_data->psk_id == NULL)
+            {
+                free(device_data->psk);
+                device_data->psk = oldpsk;
+                device_data->psk_id = oldid;
+                return -1;
+            }
+            memcpy(device_data->psk_id, binary_buffer, length);
+            free(oldpsk);
+            free(oldid);
+
+            return 0;
+        }
+    }
+
+    return -1;
+}
+
+static int rest_devices_remove_list(rest_list_t *list, const char *id)
+{
+    rest_list_entry_t *device_entry;
+    database_entry_t *device_data;
+
+    for (device_entry = list->head; device_entry != NULL; device_entry = device_entry->next)
+    {
+        device_data = (database_entry_t *)device_entry->data;
+
+        if (strcmp(id, device_data->uuid) == 0)
+        {
+            rest_list_remove(list, (void *)device_data);
+            return 0;
+        }
+    }
+
+    return -1;
+}
+
+int rest_devices_get_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void *context)
+{
+    rest_context_t *rest = (rest_context_t *)context;
+    rest_list_t *device_list = rest->devicesList;
+    char string[512];
+    size_t length;
+    database_entry_t *device_data;
+    rest_list_entry_t *device_entry;
+    json_t *jdevices = NULL;
+
+    rest_lock(rest);
+
+    jdevices = json_array();
+    for (device_entry = device_list->head; device_entry != NULL; device_entry = device_entry->next)
+    {
+        device_data = (database_entry_t *)device_entry->data;
+        length = sizeof(string);
+        memset(string, 0, length);
+
+        if (base64_encode(device_data->psk_id, device_data->psk_id_len, string, &length))
+        {
+            ulfius_set_empty_body_response(resp, 500);
+            goto exit;
+        }
+
+        json_t *jstring = json_string(string);
+        json_t *jobject = json_object();
+        json_object_set_new(jobject, "psk_id", jstring);
+        json_array_append_new(jdevices, jobject);
+    }
+
+    ulfius_set_json_body_response(resp, 200, jdevices);
+exit:
+    json_decref(jdevices);
+    rest_unlock(rest);
+
+    return U_CALLBACK_COMPLETE;
+}
+
+int rest_devices_get_name_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void *context)
+{
+    rest_context_t *rest = (rest_context_t *)context;
+    rest_list_t *device_list = rest->devicesList;
+    char string[512];
+    size_t length = sizeof(string);
```

Comment:
Should multiply by type to get accurate data length.

Reply:
sizeof returns size of the object (memory), not just the length of the array.

Reply:
oops, my bad.